File: statcheck-ml/r/normalize.R
# Make the text look the same whatever engine read the PDF.
#
# This is the R port of `statcheck_ml/normalize.py`. It reads the same
# `spec/normalize.json`, so the rules exist once and this file only applies
# them. Never write a rule here that is not in the spec.
#
# R reads PDF files with pdftools, which bundles poppler. Poppler returns a
# whole column as one line, and every prefilter rule measures one line. Without
# this stage the R port keeps 0.793 of the holdout results against 0.929 for the
# Python port. With it the R port keeps 0.879.

library(jsonlite)

#' Load the shared normalisation spec.
#'
#' @param path Path to `spec/normalize.json`.
#' @param charmap_path Path to `spec/charmap.json`. Every character the model
#'   can read. A character in it is never renamed, whatever position it stands
#'   in, because renaming one can only move the input away from the text the
#'   model was trained on. Without it the stage renames the footnote marker, the
#'   multiplication sign and the significance star, and costs recall.
load_normalize_spec <- function(path, charmap_path = NULL) {
  spec <- jsonlite::fromJSON(path, simplifyVector = TRUE)
  spec$keep <- unique(c(
    strsplit(spec$text_characters, "")[[1]],
    strsplit(spec$protected_symbols, "")[[1]]
  ))
  if (is.null(charmap_path)) {
    charmap_path <- file.path(dirname(path), "charmap.json")
  }
  spec$known <- if (file.exists(charmap_path)) {
    names(jsonlite::fromJSON(charmap_path, simplifyVector = FALSE)$chars)
  } else {
    character(0)
  }
  spec
}

#' Cut joined columns back into lines of a normal width.
#'
#' A line shorter than the trigger is never touched, so text from an engine that
#' already breaks lines per page line passes through almost unchanged.
reflow <- function(text, spec) {
  text <- gsub("\r\n", "\n", text, fixed = TRUE)
  text <- gsub("\r", "\n", text, fixed = TRUE)
  target <- spec$target_line_width
  trigger <- spec$reflow_trigger_width

  # `strsplit` drops the empty piece after a final separator and Python's
  # `split` keeps it. Without this the R port loses the last newline of every
  # document, and the three ports return texts of different length.
  lines <- strsplit(text, "\n", fixed = TRUE)[[1]]
  if (grepl("\n$", text)) lines <- c(lines, "")

  out <- character(0)
  for (line in lines) {
    while (nchar(line) > trigger) {
      # `cut` counts from zero, so that this matches Python's
      # `line.rfind(" ", target %/% 2, target)` exactly. R counts from one, and
      # mixing the two conventions keeps the space the cut lands on, which made
      # the R port disagree with the other two.
      window <- substr(line, 1, target)          # positions 1..target
      spaces <- gregexpr(" ", window, fixed = TRUE)[[1]]
      spaces <- spaces[spaces > 0] - 1           # now counted from zero
      spaces <- spaces[spaces >= target %/% 2]   # Python's search starts here
      cut <- if (length(spaces) == 0) target else max(spaces)
      out <- c(out, substr(line, 1, cut))        # Python's line[:cut]
      line <- sub("^\\s+", "", substr(line, cut + 1, nchar(line)))
    }
    out <- c(out, line)
  }
  paste(out, collapse = "\n")
}

#' Rename damaged operator characters to the alphabet the model knows.
#'
#' A character counts as a damaged operator when the model cannot read it AND it
#' stands where an operator belongs. Both halves are needed. The position alone
#' is not enough, because the copyright sign, the multiplication sign and the
#' significance star all stand between a letter and a digit, and the model reads
#' all three well.
canonicalise <- function(text, spec) {
  slots <- spec$canonical_operator_slots
  matches <- gregexpr(spec$operator_site_pattern, text, perl = TRUE)[[1]]
  if (matches[1] == -1) return(list(text = text, mapping = character(0)))

  starts <- attr(matches, "capture.start")
  lengths <- attr(matches, "capture.length")
  chars <- substring(text, starts, starts + lengths - 1)
  chars <- chars[nzchar(chars)]

  damaged <- unique(chars[!(chars %in% spec$known) & !(chars %in% spec$keep) &
                            !(chars %in% slots)])
  if (length(damaged) == 0) return(list(text = text, mapping = character(0)))

  # A slot already used in this document keeps its meaning, so the rename must
  # not take it. Inside one document the correspondence between two engines is
  # one to one, so the free slots are enough.
  used <- vapply(slots, function(s) grepl(s, text, fixed = TRUE), logical(1))
  free <- slots[!used]
  if (length(free) == 0) return(list(text = text, mapping = character(0)))

  counts <- vapply(damaged,
                   function(c) lengths(regmatches(text, gregexpr(c, text, fixed = TRUE))),
                   numeric(1))
  # The tie is broken by the code point, never by `order` on the characters
  # themselves. String order in R follows the locale, so the same input would
  # assign different slots on different machines and the ports would disagree.
  code_points <- vapply(damaged, utf8ToInt, integer(1))
  order_by_count <- damaged[order(-counts, code_points)]
  order_by_count <- head(order_by_count, length(free))

  mapping <- setNames(free[seq_along(order_by_count)], order_by_count)
  for (from in names(mapping)) {
    text <- gsub(from, mapping[[from]], text, fixed = TRUE)
  }
  list(text = text, mapping = mapping)
}

#' Apply every rule that makes the text engine independent.
normalize_text <- function(text, spec) {
  flowed <- reflow(text, spec)
  canonicalise(flowed, spec)$text
}

File: statcheck-ml/js/normalize.js
// Make the text look the same whatever engine read the PDF.
//
// This is the JavaScript port of `statcheck_ml/normalize.py`. It reads the same
// `spec/normalize.json`, so the rules exist once and this file only applies
// them. Never write a rule here that is not in the spec.
//
// A browser gets PDF.js, which returns positioned pieces rather than lines, and
// a port must join them before this stage runs. Measured on 198 holdout
// documents, PDF.js holds 0.926 of the results and keeps 0.907 after the
// prefilter, against 0.929 for PyMuPDF, which is what the models were trained
// on.

/**
 * Build a normaliser from the shared spec.
 *
 * @param {object} spec Parsed `spec/normalize.json`.
 * @param {object} [charmap] Parsed `spec/charmap.json`. Every character the
 *   model can read. A character in it is never renamed, whatever position it
 *   stands in, because renaming one can only move the input away from the text
 *   the model was trained on. Without it the stage renames the footnote marker,
 *   the multiplication sign and the significance star, and costs recall.
 */
export function createNormalizer(spec, charmap) {
  const known = new Set(charmap ? Object.keys(charmap.chars) : []);
  const targetWidth = spec.target_line_width;
  const reflowTrigger = spec.reflow_trigger_width;
  const slots = spec.canonical_operator_slots;
  const operatorSite = new RegExp(spec.operator_site_pattern, 'gu');
  const keep = new Set([
    ...spec.text_characters,
    ...spec.protected_symbols,
  ]);

  /**
   * Cut joined columns back into lines of a normal width.
   *
   * A line shorter than the trigger is never touched, so text that already
   * breaks per page line passes through almost unchanged.
   */
  function reflow(text) {
    const normalised = text.replace(/\r\n/g, '\n').replace(/\r/g, '\n');
    const out = [];
    for (let line of normalised.split('\n')) {
      while (line.length > reflowTrigger) {
        // The search window is [targetWidth / 2, targetWidth), matching
        // Python's `rfind(" ", start, end)`, whose end is exclusive. Searching
        // one character wider picks a different cut and the ports disagree.
        let cut = line.lastIndexOf(' ', targetWidth - 1);
        if (cut < Math.floor(targetWidth / 2)) cut = targetWidth;
        out.push(line.slice(0, cut));
        line = line.slice(cut).replace(/^\s+/, '');
      }
      out.push(line);
    }
    return out.join('\n');
  }

  /**
   * Rename damaged operator characters to the alphabet the model knows.
   *
   * A character counts as a damaged operator when the model cannot read it AND
   * it stands where an operator belongs. Both halves are needed. The position
   * alone is not enough, because the copyright sign, the multiplication sign
   * and the significance star all stand between a letter and a digit, and the
   * model reads all three well.
   *
   * @returns {{text: string, mapping: Map<string, string>}}
   */
  function canonicalise(text) {
    const damaged = new Set();
    operatorSite.lastIndex = 0;
    let match;
    while ((match = operatorSite.exec(text)) !== null) {
      const ch = match[1];
      if (!known.has(ch) && !keep.has(ch) && !slots.includes(ch)) damaged.add(ch);
    }
    if (damaged.size === 0) return { text, mapping: new Map() };

    // A slot already used in this document keeps its meaning, so the rename
    // must not take it. Inside one document the correspondence between two
    // engines is one to one, so the free slots are enough.
    const free = slots.filter((c) => !text.includes(c));
    if (free.length === 0) return { text, mapping: new Map() };

    const count = (c) => text.split(c).length - 1;
    const order = [...damaged]
      .sort((a, b) => count(b) - count(a) || (a < b ? -1 : 1))
      .slice(0, free.length);

    const mapping = new Map();
    order.forEach((ch, i) => mapping.set(ch, free[i]));

    let out = '';
    for (const ch of text) out += mapping.get(ch) ?? ch;
    return { text: out, mapping };
  }

  /**
   * Apply every rule that makes the text engine independent.
   */
  function normalize(text) {
    const linesBefore = text.split('\n').length;
    const flowed = reflow(text);
    const { text: fixed, mapping } = canonicalise(flowed);
    const renamed = {};
    for (const [k, v] of mapping) renamed[k] = v.codePointAt(0);
    return {
      text: fixed,
      info: {
        lines_before: linesBefore,
        lines_after: fixed.split('\n').length,
        operators_renamed: renamed,
      },
    };
  }

  return { reflow, canonicalise, normalize };
}

/**
 * Join what PDF.js returns into lines, before the normaliser runs.
 *
 * PDF.js reports text as pieces with a position and no lines. A port must
 * decide where a line ends, and the rule must match this one, or the port reads
 * a different document from the one measured.
 *
 * @param {object} textContent The value of `page.getTextContent()`.
 */
export function joinPdfJsItems(textContent) {
  const parts = [];
  let lastY = null;
  for (const item of textContent.items) {
    if (item.str === undefined) continue;
    const y = item.transform ? Math.round(item.transform[5]) : null;
    // A change of vertical position of more than two units is a new line.
    if (lastY !== null && y !== null && Math.abs(y - lastY) > 2) parts.push('\n');
    parts.push(item.str);
    if (item.hasEOL) parts.push('\n');
    lastY = y;
  }
  return parts.join('');
}
